File: haskell/Day12.hs
module Day12 where

import qualified Data.Map as M
import qualified Data.Array as A
import qualified Data.Array.IArray as IA
import Text.Read (readMaybe)
import Debug.Trace

import Utils
inputFile :: String
inputFile = "day12.txt"

type Reg = Char
type Registry = M.Map Reg Int

data Operand = Register Reg | Constant Int deriving (Show)

getValue :: Registry -> Operand -> Int
getValue _ (Constant x) = x
getValue reg (Register r) = M.findWithDefault 0 r reg

data Instruction = Copy Operand Reg
                 | Inc Reg
                 | Dec Reg
                 | Jnz Operand Int
                 deriving (Show)

parseOperand :: String -> Operand
parseOperand x = case readMaybe x :: Maybe Int of
    Just x' -> Constant x'
    Nothing -> Register $ head x

parseInstruction :: String -> Instruction
parseInstruction = f . words
    where
        f :: [String] -> Instruction
        f ["cpy", x, y] = Copy (parseOperand x) (head y)
        f ["inc", a] = Inc $ head a
        f ["dec", a] = Dec $ head a
        f ["jnz", x, y] = Jnz (parseOperand x) (read y)
        f _ = error "This should not happen"

type Program = A.Array Int Instruction

data ProgramState = State
    { program :: Program
    , pc :: Int
    , registry :: Registry
    , finished :: Bool
    } deriving Show

emptyState1 :: Program -> ProgramState
emptyState1 p = State { program = p, pc = 0, registry = M.empty, finished = False }

emptyState2 :: Program -> ProgramState
emptyState2 p = State { program = p, pc = 0, registry = M.fromList [('c', 1)], finished = False }

getInstruction :: Program -> Int -> Maybe Instruction
getInstruction p i = let
    (l, u) = IA.bounds p
    in
        if l <= i  && i <= u
        then Just $ p IA.! i
        else Nothing

runNextInstruction :: ProgramState -> ProgramState
runNextInstruction s@State { program = p, pc = pc, registry = reg } =
    case getInstruction p pc of
        Nothing -> s { finished = True}
        Just i -> case i of
            Copy x r -> s { pc = pc+1, registry = M.insert r (getValue reg x) reg }
            Inc r -> s { pc = pc+1, registry = M.insertWith (+) r 1 reg }
            Dec r -> s { pc = pc+1, registry = M.insertWith (+) r (-1) reg }
            Jnz n i -> if getValue reg n /= 0
                then s { pc = pc+i }
                else s { pc = pc+1 }

iterateWhile :: (a -> Bool) -> (a -> a) -> a -> a
iterateWhile c f = head . dropWhile c . iterate f

part :: (Program -> ProgramState) -> String -> Int
part e x = let
    is = (map parseInstruction . lines) x
    program = IA.listArray (0, length is - 1) is
    initState = e program
    states = iterate runNextInstruction initState
    finalState = iterateWhile (not . finished) runNextInstruction initState
    in registry finalState M.! 'a'

part1 :: String -> Int
part1 = part emptyState1

part2 :: String -> Int
part2 = part emptyState2

File: rust/src/day12.rs
use crate::common::Day;

pub struct Day12;

type Register = u8;
type Value = i32;
type Registry = Vec::<i32>;
type Program = Vec::<Instruction>;

#[derive(Debug)]
pub enum Operand {
    Const(i32),
    Reg(Register)
}

impl Operand {
    fn get_value(&self, registry: &Registry) -> Value{
        match self {
            Self::Const(x) => *x,
            Self::Reg(r) => registry[*r as usize]
        }
    }
}

#[derive(Debug)]
pub enum Instruction {
    Copy(Operand, Register), 
    Inc(Register),
    Dec(Register),
    Jnz(Operand, Value),
}

#[derive(Debug)]
struct ProgramState {
    program: Program,
    pc: usize,
    registry: Registry
}

fn parse_operand(input: &str) -> Operand {
    match input.parse::<i32>() {
        Ok(x) => Operand::Const(x),
        Err(_) => {
            let reg = parse_reg(input);
            return Operand::Reg(reg)
        }
    }
}

fn parse_reg(input: &str) -> Register {
    match input {
        "a" => 0,
        "b" => 1,
        "c" => 2,
        "d" => 3,
        _ => panic!("Unexpected register: {}", input),
    }
}

fn parse_value(input: &str) -> Value {
    input.parse::<i32>().expect("Value is not a valid int")
}

fn parse_instruction(input: &str) -> Instruction {
    let words: Vec::<&str> = input.split(' ').collect();
    match words[0] {
        "cpy" => Instruction::Copy(parse_operand(words[1]), parse_reg(words[2])),
        "inc" => Instruction::Inc(parse_reg(words[1])),
        "dec" => Instruction::Dec(parse_reg(words[1])),
        "jnz" => Instruction::Jnz(parse_operand(words[1]), parse_value(words[2])),
        _ => panic!("Unexpected instruction")
    }
}

fn step(program_state: &mut ProgramState) -> bool {
    let instruction = &program_state.program[program_state.pc];
    match instruction {
        Instruction::Copy(op, reg) => {
            program_state.pc += 1;
            program_state.registry[reg] = op.get_value(program_state.registry);
        },
        Instruction::Inc(r) => {
            program_state.pc += 1;
            program_state.registry[r] += 1;
        },
        Instruction::Dec(r) => {
            program_state.pc += 1;
            program_state.registry[r] -= 1;
        },
        Instruction::Jnz(op, val) => {
            program_state.pc = program_state.pc.checked_add(val);
        }
    }
}

fn run_program(program: Program) {
    let mut state = ProgramState {
        program,
        pc: 0,
        registry: vec![0, 0, 0, 0]
    };

    while step(&mut state) {}
}

impl Day for Day12 {
    type Input = Vec<Instruction>;
    type Output = i32;

    fn get_input_file() -> &'static str {
        "day12"
    }
    
    fn parse(input: &str) -> Self::Input {
        input.split('\n').map(parse_instruction).collect()
    }

    fn part1(input: &Self::Input) -> Self::Output {
        println!("{:?}", input);
        0
    }

    fn part2(input: &Self::Input) -> Self::Output {
        0
    }
}